Improve getSpecificPrice static cache for products with combinations

Sollsy opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project!

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/#pull-requests

For type and category see:
https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/pull-requests/#type--category
------------------------------------------------------------------------------>

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 9.0.x
| Description?      | getSpecificPrice is called on every combination for product and caching one result at a time. Instead we should call it once for all product variants and cache them. This way we reduce memory usage, query doubles and loading speed. Specific prices is allways returned with results from query or false if no match found. Tested on 8.2.2 and 8.1.7. The more combinations product has the bigger improvement you will notice.
| Type?             | improvement
| Category?         | FO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | create product with a lot of combinations 50-100+. Below you can find improvements on product with 800 combinations. 
| UI Tests          |  CI test are green
| Fixed issue or discussion?     | 
| Related PRs       | 
| Sponsor company   | 

Should work on higher version but i dont have 
Below you can see difference in loading product page with 800 combinations, which looks like that from customer point of view.
<img width="623" height="328" alt="image" src="https://github.com/user-attachments/assets/4ba0b21b-4301-48e6-8ba6-9feac41aaeaa" />
Before change:
<img width="1135" height="611" alt="image" src="https://github.com/user-attachments/assets/3f4aedb6-89b7-4c8b-bb9f-13cb12d62b9f" />

After change:
<img width="1103" height="581" alt="image" src="https://github.com/user-attachments/assets/f2c93033-a7db-4533-a8a2-d2f8229961c1" />

In combination with another change improving caching of StockAvailable::getQuantityAvailableByProduct, we can go even lower with memory usage and loading speed.
before StockAvailable cache change
<img width="1249" height="603" alt="image" src="https://github.com/user-attachments/assets/17ac6ff8-574a-42c8-add3-1da3c78807ba" />
After StockAvailable cache change
<img width="1251" height="607" alt="image" src="https://github.com/user-attachments/assets/b70f48af-8ded-4e4d-8dca-e4bcf4d01a73" />

Sollsy committed:
better combinations caching for getSpecificPrice

Improvement on checking specific prices for products with a lot of combinations.

ps-jarvis commented:
<!-- PR_WELCOME -->
Hello @Sollsy!

This is your first pull request on [PrestaShop](https://github.com/PrestaShop/PrestaShop/) repository of the PrestaShop project.

Thank you, and welcome to this Open Source community!

Sollsy committed:
Improve static cache for product combinations StockAvailable.php

Hlavtox reviewed (changes requested):
Hi @Sollsy, this looks very promising! 🥳

But to merge it, we must improve it a bit:
- Please make a separate pr for the stock optimizazion, it's not related. For clarity.
- I see some string replacements in the query, please avoid this and construct the query cleanly.

Also, please make absolutely sure that these caches are wiped after modification, so we avoid potential bugs in one request.

Sollsy committed:
Update StockAvailable.php

Sollsy committed:
change to query

keep old query as a fallback for old method and add $queryAll and $query_extraAll to cache all variants specific price

Sollsy committed:
Update SpecificPrice.php

Fill all variants only when id_product_attribute was provided to prevent table stress when id_product_attribute is 0;
